contracts/src/ProtocolAdapter.sol

```diff
@@ -291,16 +289,56 @@ contract ProtocolAdapter is IProtocolAdapter, ReentrancyGuardTransient, Commitme
 
     /// @notice An internal function verifying a RISC0 logic proof.
     /// @param input The verifier input of the logic proof.
+    /// @param root The root of the action tree containing all tags of an action.
+    /// @param consumed Bool indicating whether the tag is a commitment or a nullifier.
     /// @dev This function is virtual to allow for it to be overridden, e.g., to mock proofs with a mock verifier.
-    function _verifyLogicProof(Logic.VerifierInput calldata input) internal view virtual {
+    function _verifyLogicProof(Logic.VerifierInput calldata input, bytes32 root, bool consumed) internal view virtual {
         // slither-disable-next-line calls-loop
         _TRUSTED_RISC_ZERO_VERIFIER_ROUTER.verify({
             seal: input.proof,
             imageId: input.verifyingKey,
-            journalDigest: input.instance.toJournalDigest()
+            journalDigest: input.toJournalDigest(root, consumed)
         });
     }
 
+    /// @notice Verifies the forwarder calls of a given action.
+    /// @param input The input to verify the forwarder calls for.
+    /// @param consumed The bool indicating whether the tag is commitment of nullifier
+    function _verifyForwarderCall(Logic.VerifierInput calldata input, bool consumed) internal view {
+        // The PA expects the forwarder calldata to be present at the head of the external payload
+        ForwarderCalldata memory call = abi.decode(input.appData.externalPayload[0].blob, (ForwarderCalldata));
+        // The plaintext should be stored at the head of resource payload and be decodable
+        Resource memory resource = abi.decode(input.appData.resourcePayload[0].blob, (Resource));
+        // slither-disable-next-line calls-loop
+        bytes32 fetchedKind = IForwarder(call.untrustedForwarder).calldataCarrierResourceKind();
+
+        // Check kind correspondence
+        if (resource.kind() != fetchedKind) {
+            revert CalldataCarrierKindMismatch({expected: fetchedKind, actual: resource.kind()});
+        }
+
+        // Check tag correspondence
+        if (!consumed) {
+            // If created, just commit the plaintext and check agains the tag
+            if (resource.commitment() != input.tag) {
+                revert CalldataCarrierCmMismatch({
+                    resource: resource,
+                    actual: input.tag,
+                    expected: resource.commitment()
+                });
+            }
+        } else if (
+            // If consumed, we expect the nullifier key to be present in the resource payload as well
```

Comment:
Comment doesn't appear to match the code, what does this have to do with the "nullifier key"?

Reply:
I see